docs/development/profiling/README.md

```diff
@@ -0,0 +1,62 @@
+# Application profiling
+
+In case the application performs poorly when serving a request, profiling can help to spot the areas in the code that need improving. This guide explains how to start the application in a way that profiling tools are run. It currently focuses on the backend especially on using [flamegraph](https://github.com/SamSaffron/flamegraph).
+
+## Adding `thin` to the gemfile
+
+The default web server OpenProject runs on is [puma](https://puma.io/), which is threaded. The profiling tool used is [rack-mini-profiler](https://github.com/MiniProfiler/rack-mini-profiler) which makes use of i.e. [stackprof](https://github.com/tmm1/stackprof). Those tools cannot work in a threaded environment which is why [thin](https://github.com/macournoyer/thin) needs to be employed. This is achieved by adding
+
+```
+gem 'thin'
+```
```

Comment:
Do we really need thin, or would running puma in single worker mode also suffice?

``
`
OPENPROJECT_WEB_WORKERS=0 OPENPROJECT_WEB_MIN__THREADS=1 OPENPROJECT_WEB_MAX__THREADS=1 RAILS_ENV=development rails s
```

Reply:
I tried running it in single worker mode but under the hood, `puma` still seems to be relying on some `Thread` mechanism. I did not dig into it further but the flamegraphs always end with a call to some method on `Thread` which I don't remember now. Using `thin` eliminated the problem.

Reply:
Thanks for the clarification! 👍